### PR TITLE
Adding --tls option to domain create command

### LINKED ISSUE #1408: Support `tls` option for `kn domain create`
### Feature request

- DomainMapping supports `spec.tls` field by https://github.com/knative/serving/pull/11250.
- I would like to request `kn domain create` supports an option to fill in the `spec.tls` field.

### Use case

- As an Knative user, I would like to create DomainMapping with custom certificate by kn command so that I can generate the DomainMapping by only one command.

### UI Example

`kn domain create` supports `--tls` option. For example, the following command:

```
$ kn domain create hello-example.default.example.org --tls=my-certs --ref=hello-example
```

creates this DomainMapping:

```
apiVersion: serving.knative.dev/v1beta1
kind: DomainMapping
metadata:
  annotations:
  name: hello-example.default.example.org
  namespace: default
spec:
  ref:
    apiVersion: serving.knative.dev/v1
    kind: Service
    name: hello-example
    namespace: default
  tls:
    secretName: my-certs
```

/kind proposal
/cc  @dsimansk 
FYI @shinigambit @julz

### PR DESCRIPTION
**Pull request by vyasgun**
## Description
Support for TLS optionn in domain create using `--tls` flag

## Changes

* Added the new flag
* Added unit tests for the new flag

## Reference

Fixes #1408

### CODE REVIEW

File: CHANGELOG.adoc [hunk 1]
```diff
@@ -17,6 +17,10 @@
 |===
 | | Description | PR
 
+| 🎁
+| Adding --tls option to domain create command
```
Comment (dsimansk) [inline CHANGELOG.adoc:21]:
```suggestion
| Adding `--tls` option to domain create command
```

File: pkg/kn/commands/domain/create.go [hunk 2]
```diff
@@ -32,8 +33,8 @@ func NewDomainMappingCreateCommand(p *commands.KnParams) *cobra.Command {
 		Use:   "create NAME",
 		Short: "Create a domain mapping",
 		Example: `
-  # Create a domain mappings 'hello.example.com' for Knative service 'hello'
-  kn domain create hello.example.com --ref hello`,
+  # Create a domain mappings 'hello.example.com' for Knative service 'hello' with TLS secret set
+  kn domain create hello.example.com --ref hello --tls tls-secret`,
```
Comment (dsimansk) [inline pkg/kn/commands/domain/create.go:37]: Or maybe `tls-secret-name`, but let's present it like a user owned input.
```suggestion
  kn domain create hello.example.com --ref hello --tls my-tls-secret`,
```

File: pkg/kn/commands/domain/create.go [hunk 3]
```diff
@@ -53,9 +54,14 @@ func NewDomainMappingCreateCommand(p *commands.KnParams) *cobra.Command {
 				return err
 			}
 
+			errs := validation.IsDNS1123Subdomain(refFlags.tls)
```
Comment (dsimansk) [inline pkg/kn/commands/domain/create.go:57]: I wonder if we shouldn't validate `name` param similarly.

Comment (dsimansk) [reply pkg/kn/commands/domain/create.go:57]: Nevermind that's probably covered by webhook once submitted.

Comment (vyasgun) [reply pkg/kn/commands/domain/create.go:57]: Yes, it is covered by webhook :smile:

**Comment by dsimansk** [comment]
Thanks @vyasgun!
/lgtm

**Review by maximilien — COMMENTED** [review]
Thanks for the contribution
/ok-to-test

File: docs/cmd/kn_domain_create.md [hunk 4]
```diff
@@ -10,8 +10,8 @@ kn domain create NAME
 
 ```
 
-  # Create a domain mappings 'hello.example.com' for Knative service 'hello'
-  kn domain create hello.example.com --ref hello
+  # Create a domain mappings 'hello.example.com' for Knative service 'hello' with TLS secret set
+  kn domain create hello.example.com --ref hello --tls my-tls-secret
```
Comment (rhuss) [inline docs/cmd/kn_domain_create.md:14]: could we call `my-tls-secret` --> `my-cert-secret` ? (as TLS is more the overall mechanism that is enabled by this option, but the secret holds the server certificate used during TLS communication, so we can be more specific here)

File: docs/cmd/kn_domain_create.md [hunk 5]
```diff
@@ -10,8 +10,8 @@ kn domain create NAME
 
 ```
 
-  # Create a domain mappings 'hello.example.com' for Knative service 'hello'
-  kn domain create hello.example.com --ref hello
+  # Create a domain mappings 'hello.example.com' for Knative service 'hello' with TLS secret set
```
Comment (rhuss) [inline docs/cmd/kn_domain_create.md:13]:
```suggestion
  # Create a domain mappings 'hello.example.com' for Knative service 'hello' pointing to the secret that holds the server certificate
```

File: docs/cmd/kn_domain_create.md [hunk 6]
```diff
@@ -20,6 +20,7 @@ kn domain create NAME
   -h, --help               help for create
   -n, --namespace string   Specify the namespace to operate in.
       --ref string         Addressable target reference for Domain Mapping. You can specify a Knative service, a Knative route. Examples: '--ref' ksvc:hello' or simply '--ref hello' for a Knative service 'hello', '--ref' kroute:hello' for a Knative route 'hello'. '--ref ksvc:mysvc:mynamespace' for a Knative service 'mysvc' in another namespace 'mynamespace', If a prefix is not provided, it is considered as a Knative service in the current namespace. If referring to a Knative service in another namespace, 'ksvc:name:namespace' combination must be provided explicitly.
+      --tls string         TLS secret name.
```
Comment (rhuss) [inline docs/cmd/kn_domain_create.md:23]:
```suggestion
      --tls string         Enable TLS and point to the secret that holds the server certificate.
```

File: pkg/kn/commands/domain/create.go [hunk 7]
```diff
@@ -32,8 +33,8 @@ func NewDomainMappingCreateCommand(p *commands.KnParams) *cobra.Command {
 		Use:   "create NAME",
 		Short: "Create a domain mapping",
 		Example: `
-  # Create a domain mappings 'hello.example.com' for Knative service 'hello'
-  kn domain create hello.example.com --ref hello`,
+  # Create a domain mappings 'hello.example.com' for Knative service 'hello' with TLS secret set
+  kn domain create hello.example.com --ref hello --tls my-tls-secret`,
```
Comment (rhuss) [inline pkg/kn/commands/domain/create.go:37]:
```suggestion
  kn domain create hello.example.com --ref hello --tls my-cert-secret`,
```

File: pkg/kn/commands/domain/create.go [hunk 8]
```diff
@@ -32,8 +33,8 @@ func NewDomainMappingCreateCommand(p *commands.KnParams) *cobra.Command {
 		Use:   "create NAME",
 		Short: "Create a domain mapping",
 		Example: `
-  # Create a domain mappings 'hello.example.com' for Knative service 'hello'
-  kn domain create hello.example.com --ref hello`,
+  # Create a domain mappings 'hello.example.com' for Knative service 'hello' with TLS secret set
```
Comment (rhuss) [inline pkg/kn/commands/domain/create.go:36]:
```suggestion
  # Create a domain mappings 'hello.example.com' for Knative service 'hello' pointing to the secret that holds the server certificate
```

File: pkg/kn/commands/domain/domain.go [hunk 9]
```diff
@@ -44,6 +44,7 @@ func NewDomainCommand(p *commands.KnParams) *cobra.Command {
 
 type RefFlags struct {
 	reference string
+	tls       string
```
Comment (rhuss) [inline pkg/kn/commands/domain/domain.go:47]: happy to keep it, but maybe `certSecret` would be more appropriate ?

File: pkg/serving/v1alpha1/client.go [hunk 10]
```diff
@@ -160,6 +160,15 @@ func (b *DomainMappingBuilder) Reference(reference duckv1.KReference) *DomainMap
 	return b
 }
 
+// TLS for domainMapping builder
+func (b *DomainMappingBuilder) TLS(tls string) *DomainMappingBuilder {
```
Comment (rhuss) [inline pkg/serving/v1alpha1/client.go:164]:
```suggestion
func (b *DomainMappingBuilder) TLS(cert string) *DomainMappingBuilder {
```

**Review by rhuss — COMMENTED** [review]
Thanks, looks good to me ! I have some minor proposal wrt/ to naming, maybe we could change this so that we can add this PR to the upcoming release 0.25 (due today :) ?

File: pkg/kn/commands/domain/domain.go (hunk 9 above)
Comment (vyasgun) [reply pkg/kn/commands/domain/domain.go:47]: I think `tls` might be better because it corresponds to the flag name (`--tls`).

**Review by rhuss — APPROVED** [review]
Thanks !

/approve
/lgtm

**Comment by rhuss** [comment]
/retest

Looks like there is some flake in the e2e tests, which might be due to recreating the same domain too fast after it has been deleted. @vyasgun good we adjust the test to use a different domain for the `--tls` test, so that we can avoid the race condition in the future ?

```
      ┣━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
        🦆 kn domain delete hello.example.com --namespace kne2etests12
        ┃ Domain mapping 'hello.example.com' deleted in namespace 'kne2etests12'.
        ┃ 
        ┣━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
        🦆 kn domain create hello.example.com --ref hello --tls tls-secret --namespace kne2etests12
        ┃ 
        🔥 Error: object is being deleted: domainmappings.serving.knative.dev "hello.example.com" already exists
        🔥 Run 'kn --help' for usage
        🔥 
        =================================================================
```

**Comment by dsimansk** [comment]
Thanks!

/approve
/lgtm

**Comment by rhuss** [comment]
The new error might be also a race condition, e.g. `describe` too quick after the domain has been created. I wonder whether we should also add a sync-wait when creating the domain in a future enhancement. For now, I would try to add some hardcoded delay between `create` and `describe`.

Also, I wonder whether the serving controller checks, if the referenced secret really exists (not sure if you have created in the test, too ?)

**Comment by dsimansk** [comment]
And maybe it's working well only for `example.com`? I wonder if `newdomain.com` shouldn't be rather e.g. `new.example.com`.

**Comment by vyasgun** [comment]
@rhuss  no it doesn't check if the secret exists. As per the referenced PR for the serving side change, the secret should be created if it doesn't exist. I haven't tested what happens in the backend but mentioning a TLS secret name in the spec changes the URL to https

**Pull request merged**